### DISCUSSION TITLE
Is it possible to pass state from component to Custom Document?

### DISCUSSION THREAD

**Discussion by AlexTechNoir** [vercel/next.js#23311] [Help]
Good day!

I'm implementing cookie consent with Google Analytics and I need to toggle value inside "script" tag in Head of Custom Document based on state that located in Layout.js component. 

In **_document.js** in Head I have a script tag that I took from [official example](https://github.com/vercel/next.js/blob/canary/examples/with-google-analytics/pages/_document.js#L15):
```
...
<script
  dangerouslySetInnerHTML={{
    __html: `
      window['ga-disable-${GA_TRACKING_ID}'] = true    <-- I need to toggle this value on consent
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());
      gtag('config', '${GA_TRACKING_ID}', {
        page_path: window.location.pathname,
      });
    `,
  }}
/>
...
```
Is there a way to get state from component to Custom Document?

**Comment by JacobLey** [comment]
I've found a way that works. Note it uses some NextJS internals so not future proof, but should help.
It loads the value found in `_NEXT_DATA_`. If you are attaching this data somewhere in the request (e.g. returning it in getServerSideProps) then it should be somewhere in this object

Try something like this for a custom document:
```
import Document, { Html, Head, Main, NextScript } from 'next/document.js';
import { DocumentContext } from 'next/dist/next-server/lib/document-context.js';
import { useContext } from 'react';

const GoogleScript = () => {

    const nextData = useContext(DocumentContext).__NEXT_DATA__;

    return <script dangerouslySetInnerHTML={<custom script with data from nextData>}/>
};

export default class CustomDocument extends Document {

    public render() {
        return (
            <Html>
                <Head>
                    <GoogleScript/>
                </Head>
                <body>
                    <Main/>
                    <NextScript/>
                </body>
            </Html>
        );
    }
}
```

Idk if you are using things like amp, but if you are you should check that value (also found in the context) before adding any custom scripts

**Reply by AlexTechNoir** [reply]
I see you useContext here. Is it possible to pass usual data though it to Custom Document? I'd really prefer future proof approach to not to get back to the same implementation many times.

**Reply by JacobLey** [reply]
What do you mean? The response of `useContext` is described here https://github.com/vercel/next.js/blob/canary/packages/next/next-server/lib/utils.ts#L173 so all of that is available in CustomDocument. With the `__NEXT_DATA__` being the stuff explicitly returned by intial/server props.

You could pull that entire render function out as a functional component so that you can use the `useContext` hook at the top level. Then you can pass that data to whatever elements you want.

for example:
```
const DocumentRender = () => {
    const documentContext = useContext(DocumentContext);
    return (
        <Html >
            <Head>
                <GoogleScript shouldDisable={documentContext.__NEXT_DATA__.pageProps.shouldDisableGA}/>
            </Head>
            <body>
                <Main/>
                <NextScript/>
            </body>
        </Html>
    );
};

export default class CustomDocument extends Document {

    public render() {
        return (
            <DocumentRender/>
        );
    }
}
```

**Reply by AlexTechNoir** [reply]
I'm beginning to understand now. No, I'm not attaching the data from any Next.js functions. It's just usual context state, like: if user ignored cookie banner - state initially false, cookies aren't loaded, if accepted - state=true, localStorage - 'true' string value, cookies are loaded, if rejected - state=false, localStorage - 'false' string value, cookies aren't loaded and deleted if present previously.

I tried to do something like this, but it says that context is undefined:
```
import Document, { Html, Head, Main, NextScript } from 'next/document'
import { ServerStyleSheet } from 'styled-components'
import { GA_TRACKING_ID } from '../../lib/gtag'
import { useContext, useEffect } from 'react'
import Context from '../context'

const GAScript = () => {
  const { areCookiesAccepted } = useContext(Context)
  const [ value, setValue ] = useState({
    __html: `
      window['ga-disable-${GA_TRACKING_ID}'] = ${!areCookiesAccepted}
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());
      gtag('config', '${GA_TRACKING_ID}', {
        page_path: window.location.pathname,
      });
    `,
  })

  useEffect(() => {
    setValue(value)
  }, [areCookiesAccepted])

  return (
    <script dangerouslySetInnerHTML={value} />
  )
}

class MyDocument extends Document {
  static async getInitialProps(ctx) {
    ...
  }

  render() {
    return (
      <Html lang="en">
        <Head>
          {/* Global Site Tag (gtag.js) - Google Analytics */}
          <script
            async
            src={`https://www.googletagmanager.com/gtag/js?id=${GA_TRACKING_ID}`}
          />
          <GAScript />
        </Head>
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    )
  }
}
```

**Reply by JacobLey** [reply]
Well what does `../context` look like? You have to set the value with a Provider before it can be consumed.

I would consider Document as your entrypoint so it is hard to use your own provider. I would instead attach data to `props` of a page, or potentially the `req` if you are using a custom server.

Also the document is rendered _only_ server side, so hooks like `useState` and `useEffect` has no impact

**Reply by AlexTechNoir** [reply]
Context works fine everywhere except Custom Document. Provider is in Custom App. I don't and can't use custom server.

If useEffect cannot be used in CD then it's all useless. The hook is vital for toggling.

Well, I guess, I can think of 2 not so ideal options:

1.Deleting loaded cookies every time when user ignores or rejects, which is a little bit GDPR not compliant.
2.Inserting the script not in Custom Document, but in Head tag of EVERY page, which is a bad choice in DRY vs GDPR.

Anyway, thank you for your time, Jacob!

**Reply by JacobLey** [reply]
The Document contains the App, not vice versa.

So the provider would also have to be in the Document.

Also Document is _always_ server side, while App is _sometimes_ server side. So you can't reliably use the provider in App.
And using a provider + context in the same component seems like a major antipattern.

You should extend the `getInitialProps` of the custom app to include extra props like "should enable analytics", and then fetch that data using the nextjs document context.

> useEffect cannot be used in CD then it's all useless

If you want to toggle something, you want to persist it. And server side render is a one-and-done process. There is nothing to persist.
So either you need to "persist" with something like cookies on the request or client side.

**Reply by AlexTechNoir** [reply]
Yeah, I mean, the Provider itself is in the separate file, it's just imported in App :)

I'm persisting everything in localStorage and putting it in condition in useEffect to load/delete cookies, like this:
```
  const router = useRouter()
  useEffect(() => {

    if (localStorage.getItem('areCookiesAccepted') === 'true') {
      console.log(1)
      const handleRouteChange = url => {
        gtag.pageview(url)
      }
      router.events.on('routeChangeComplete', handleRouteChange)
      return () => {
        router.events.off('routeChangeComplete', handleRouteChange)
      }
    } else if (localStorage.getItem('areCookiesAccepted') === 'false' ||
    localStorage.getItem('areCookiesAccepted') === null) {
      console.log(2)
      window[`ga-disable-${GA_TRACKING_ID}`] = true

      document.cookie = '_ga=; Max-Age=0;'

      const cookiePair = document.cookie.split('; ').find(row => row.startsWith('_ga_'))
      if (cookiePair !== undefined) {
        const cookieName = cookiePair.substring(0, cookiePair.indexOf('='))
        document.cookie = `${cookieName}=; Max-Age=0;`
      }
      
      document.cookie = '_gid=; Max-Age=0;'
    }
  }, [router.events, areCookiesAccepted])
```
 But LS value cannot be used as a dependency in useEffect, so I have a state for that. State is just a toggler basically.
